src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const clonedStates = [{ ...state }];
+
+  for (const action of actions) {
+    const newState = { ...clonedStates[clonedStates.length - 1] };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+      case 'clear':
```

Comment:
Not critical, but would be great to use variables instead of the hardcoded values